ddl/index.go

```diff
@@ -966,112 +984,87 @@ func closeAddIndexWorkers(workers []*addIndexWorker) {
 	}
 }
 
-func (w *worker) waitTaskResults(workers []*addIndexWorker, taskCnt int, totalAddedCount *int64, startHandle int64) (int64, int64, error) {
-	var (
-		addedCount int64
-		nextHandle = startHandle
-		firstErr   error
-	)
-	for i := 0; i < taskCnt; i++ {
-		worker := workers[i]
-		result := <-worker.resultCh
-		if firstErr == nil && result.err != nil {
-			firstErr = result.err
-			// We should wait all working workers exits, any way.
-			continue
+func (w *worker) waitTaskFinish(cancel context.CancelFunc, reorgInfo *reorgInfo, availableWorkerCh chan *addIndexWorker, doingTaskCh chan *reorgIndexTask, errCh chan error) error {
+	startTime := time.Now()
+	totalAddedCount := reorgInfo.Job.GetRowCount()
+	for task := range doingTaskCh {
+		select {
+		case err := <-errCh:
+			cancel()
+			return errors.Trace(err)
+		default:
 		}
-
-		if result.err != nil {
-			log.Warnf("[ddl-reorg] worker[%v] return err:%v", i, result.err)
+		task.wg.Wait()
+		// task.result should never be nil here.
+		if task.result == nil {
+			cancel()
+			return errors.Errorf("[ddl-reorg] backfill index task result is nil")
```

Comment:
Error info should not contain `[ddl-reorg]`, the logger will choose to print it.